### PR TITLE
[COLLECTIONS-858] CartesianProductIterator

### PR DESCRIPTION
**Pull request by alexey-pelykh**
The `CartesianProductIterator` is suitable to enhance a deeply-nested for-loops over a set of collections.

Given
```java
var a = new Arrays.asList('a1', 'a2', 'a3');
var b = new Arrays.asList('b1', 'b2', 'b3');
var c = new Arrays.asList('c1', 'c2', 'c3');
var d = new Arrays.asList('d1', 'd2', 'd3');
var e = new Arrays.asList('e1', 'e2', 'e3');
var f = new Arrays.asList('f1', 'f2', 'f3');
```

instead of

```java
for (var aElement : a) {
    for (var bElement : b) {
        for (var cElement : c) {
            for (var dElement : d) {
                for (var eElement : e) {
                    for (var fElement : f) {
                        // ...
                    }
                }
            }
        }
    }
}
```

it would look like:

```
var it = new CartesianProductIterator<>(a, b, c, d, e, f);
while (it.hasNext()) {
    var elements = it.next();
}
```

Extremely useful for nesting levels 10+.

### CODE REVIEW

File: src/test/java/org/apache/commons/collections4/iterators/CartesianProductIteratorTest.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    public void testEmptyCollection() {
+        final CartesianProductIterator<Character> it = new CartesianProductIterator<>(letters, empty);
+        assertFalse(it.hasNext());
+    }
+
+    /**
+     * test checking that all the tuples are returned
+     */
+    @Test
+    public void testCartesianProductExhaustivity() {
```
Comment (aherbert) [inline src/test/java/org/apache/commons/collections4/iterators/CartesianProductIteratorTest.java:79]: The test for the total output length should be supplemented with a test that the output is the desired Cartesian product:
```
[A, 1]
[A, 2]
[A, 3]
[B, 1]
[B, 2]
[B, 3]
[C, 1]
[C, 2]
[C, 3]
```

File: src/test/java/org/apache/commons/collections4/iterators/CartesianProductIteratorTest.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    @Override
+    public boolean supportsRemove() {
+        return false;
+    }
+
+    @Test
+    public void testEmptyCollection() {
+        final CartesianProductIterator<Character> it = new CartesianProductIterator<>(letters, empty);
+        assertFalse(it.hasNext());
+    }
```
Comment (aherbert) [inline src/test/java/org/apache/commons/collections4/iterators/CartesianProductIteratorTest.java:73]: This should test `next()` throws an exception.

File: src/test/java/org/apache/commons/collections4/iterators/CartesianProductIteratorTest.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import static org.junit.jupiter.api.Assertions.*;
+
+/**
+ * Test class for CartesianProductIterator.
+ */
+public class CartesianProductIteratorTest extends AbstractIteratorTest<List<Character>> {
+
+    protected List<Character> letters;
+    protected List<Character> numbers;
+    protected List<Character> empty;
```
Comment (aherbert) [inline src/test/java/org/apache/commons/collections4/iterators/CartesianProductIteratorTest.java:35]: Since empty is not reused it does not merit being created for each test fixture. You can use `Collections.emptyList()` instead; the other members can be private.

**Review by aherbert — CHANGES_REQUESTED** [review]

File: src/test/java/org/apache/commons/collections4/iterators/CartesianProductIteratorTest.java (hunk 1 above)
Comment (aherbert) [reply src/test/java/org/apache/commons/collections4/iterators/CartesianProductIteratorTest.java:79]: Since this is meant to be like a nested for loop, let's test it like one:
```java
        final CartesianProductIterator<Character> it = makeObject();
        for (final Character a : letters) {
            for (final Character b : numbers) {
                Assertions.assertEquals(Arrays.asList(a, b), it.next());
            }
        }
        assertThrows(NoSuchElementException.class, () -> it.next());
```
You could even add another test with a third iterable to have a triple loop to demonstrate it works for increasing depth.

File: src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+import java.util.*;
+
+/**
+ * This iterator creates a Cartesian product of the input enumerables,
+ * equivalent to nested for-loops.
+ * <p>
+ * The {@code remove()} operation is not supported, and will throw an
+ * {@code UnsupportedOperationException}.
+ *
```
Comment (aherbert) [inline src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java:27]: This class could do with some javadoc to state the expected output behaviour, e.g. the iterator creates the Cartesian product of the iterables as a list. Iterables provided to the constructor are used in reverse order, each until exhaustion before proceeding to the next element of the prior iterable and repeating:
```
iterable1 = A, B, C
iterable2 = 1, 2, 3
output:
A, 1
A, 2
A, 3
B, 1
B, 2
...
```

File: src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            previousTuple.set(iteratorIndex, iterator.next());
+        }
+        return new ArrayList<>(previousTuple);
+    }
+
+    @Override
+    public void remove() {
+        throw new UnsupportedOperationException("remove() is not supported");
+    }
+
```
Comment (aherbert) [inline src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java:123]: Remove empty line

File: src/test/java/org/apache/commons/collections4/iterators/CartesianProductIteratorTest.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+            final List<Character> tuple = it.next();
+            resultsList.add(tuple.toArray(new Character[0]));
+        }
+        assertThrows(NoSuchElementException.class, () -> it.next());
+        assertEquals(9, resultsList.size());
+        for (int i = 0; i < 9; i++) {
+            assertArrayEquals(expectedResultsList.get(i), resultsList.get(i));
+        }
+    }
+
```
Comment (aherbert) [inline src/test/java/org/apache/commons/collections4/iterators/CartesianProductIteratorTest.java:103]: Remove empty line

File: src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        if (previousTuple == null) {
+            previousTuple = new ArrayList<>(iterables.size());
+            for (final Iterator<? extends E> iterator : iterators) {
+                previousTuple.add(iterator.next());
+            }
+            return new ArrayList<>(previousTuple);
+        }
+
+        for (
```
Comment (aherbert) [inline src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java:102]: A single line for loop is in keeping with the style of the codebase:
```java
        for (int i = iterators.size() - 1; i >= 0; i--) {
```

**Review by aherbert — CHANGES_REQUESTED** [review]

File: src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+/**
+ * This iterator creates a Cartesian product of the input iterables,
+ * equivalent to nested for-loops.
+ * <p>
+ * The iterables provided to the constructor are used in reverse order, each
+ * until exhaustion before proceeding to the next element of the prior iterable
+ * and repeating. Consider the following example:
+ * <pre>
+ *     List<Character> iterable1 = Arrays.asList('A', 'B', 'C');
```
Comment (aherbert) [inline src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java:29]: I think this will not pass javadoc due to the `<>` characters. You need to include: `<pre>{@code` and `}</pre>` at the ends.

File: src/test/java/org/apache/commons/collections4/iterators/CartesianProductIteratorTest.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+import java.util.*;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.junit.jupiter.api.Assertions.assertArrayEquals;
+
+/**
+ * Test class for CartesianProductIterator.
```
Comment (aherbert) [inline src/test/java/org/apache/commons/collections4/iterators/CartesianProductIteratorTest.java:28]: Might as well put this in a `{@link ...}` tag.

**Review by aherbert — CHANGES_REQUESTED** [review]
This is looking good apart from the javadoc which will fail the build. If you use `<,>` chars you need a code tag.

**Comment by garydgregory** [comment]
@alexey-pelykh 
Make sure you run 'mvn' by itself before you push to run all build checks.

**Comment by alexey-pelykh** [comment]
@garydgregory sorry, my oversight - should be green now

**Comment by alexey-pelykh** [comment]
@garydgregory would it be possible to restart 23-ae? It's not exactly clear what is wrong

**Comment by alexey-pelykh** [comment]
What would be the next steps?

File: src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+ * C, 1
+ * C, 2
+ * C, 3
+ * </pre>
+ * <p>
+ * The {@code remove()} operation is not supported, and will throw an
+ * {@code UnsupportedOperationException}.
+ *
+ * @param <E>  the type of the objects being permuted
+ */
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java:62]: We need a Javadoc 'since tag here.

Comment (alexey-pelykh) [reply src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java:62]: I'd be glad to add that if I knew what to specify after `@since`

Comment (garydgregory) [reply src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java:62]: The pom.xml and changes.xml files say the next version will be 4.5.0.

Comment (alexey-pelykh) [reply src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java:65]: @garydgregory thanks! Done 👍

**Comment by garydgregory** [comment]
@alexey-pelykh 
This PR has zero text in its description.

**Comment by alexey-pelykh** [comment]
@garydgregory that is true, any specific requirements what to write there apart of what I've already added to it?

**Comment by garydgregory** [comment]
@alexey-pelykh 
I want to email the dev list and ask for reviews, and the first I do at least is read the description to know what it is I'm looking at... it's up to you to help reviewers...

**Comment by alexey-pelykh** [comment]
@garydgregory okies, done 👍

**Comment by alexey-pelykh** [comment]
Seems not much of a discussion :)

**Comment by alexey-pelykh** [comment]
👋 @garydgregory anything I can do to push this forward?

File: src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+     */
+    private final List<Iterator<? extends E>> iterators;
+
+    /**
+     * The previous generated tuple of elements.
+     */
+    private List<E> previousTuple;
+
+    /**
+     * Standard constructor for this class.
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java:84]: "Standard" does not mean anything IMO. I would say "Constructs a new instance that ...".

File: src/test/java/org/apache/commons/collections4/iterators/CartesianProductIteratorTest.java [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+    @Override
+    public boolean supportsRemove() {
+        return false;
+    }
+
+    @Test
+    public void testEmptyCollection() {
+        final CartesianProductIterator<Character> it = new CartesianProductIterator<>(letters, Collections.emptyList());
+        assertFalse(it.hasNext());
+        assertThrows(NoSuchElementException.class, () -> it.next());
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/collections4/iterators/CartesianProductIteratorTest.java:74]: Use a method reference for `next()`.

File: src/test/java/org/apache/commons/collections4/iterators/CartesianProductIteratorTest.java [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    /**
+     * test checking that all the tuples are returned
+     */
+    @Test
+    public void testCartesianProductExhaustivity() {
+        final List<Character[]> resultsList = new ArrayList<>();
+
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/collections4/iterators/CartesianProductIteratorTest.java:83]: Whitespace is not needed.

File: src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    /**
+     * Indicates if there are more tuples to return.
+     *
+     * @return true if there are more tuples, otherwise false
+     */
+    @Override
+    public boolean hasNext() {
+        for (final Iterator<? extends E> iterator : iterators) {
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java:112]: You can use the less verbose `return iterators.stream().anyMatch(Iterator::hasNext);`

File: src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+     */
+    private List<E> previousTuple;
+
+    /**
+     * Standard constructor for this class.
+     *
+     * @param iterables the iterables to create the Cartesian product from
+     * @throws NullPointerException if any of the iterables is null
+     */
+    public CartesianProductIterator(final Iterable<? extends E>... iterables) {
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java:89]: Since the code has null-guards, you can use `@SafeVarargs`.

File: src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+            Iterator<? extends E> iterator = iterators.get(i);
+            if (iterator.hasNext()) {
+                previousTuple.set(i, iterator.next());
+                return new ArrayList<>(previousTuple);
+            }
+            iterator = iterables.get(i).iterator();
+            iterators.set(i, iterator);
+            previousTuple.set(i, iterator.next());
+        }
+        throw new IllegalStateException();
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java:150]: Describe the state in the exception message.

File: src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    @Override
+    public void remove() {
+        throw new UnsupportedOperationException("remove() is not supported");
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java:155]: Keep the message language-neutral, like `java.util.Iterator.remove()`:
```java
throw new UnsupportedOperationException("remove");
```

File: src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+     * @throws NullPointerException if any of the iterables is null
+     */
+    public CartesianProductIterator(final Iterable<? extends E>... iterables) {
+        Objects.requireNonNull(iterables, "iterables");
+        this.iterables = Arrays.asList(iterables);
+        this.iterators = new ArrayList<>(iterables.length);
+        for (final Iterable<? extends E> iterable : iterables) {
+            Objects.requireNonNull(iterable, "iterable");
+            final Iterator<? extends E> iterator = iterable.iterator();
+            if (iterator.hasNext()) {
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java:96]: You don't need both blocks nested if you say:
```java
            if (!iterator.hasNext()) {
                iterators.clear();
                break;
            }
            iterators.add(iterator);
```

File: src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+                iterators.add(iterator);
+            } else {
+                iterators.clear();
+                break;
+            }
+        }
+    }
+
+    /**
+     * Indicates if there are more tuples to return.
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java:106]: "Indicates" is not what you normally see in Javadocs. When in doubt, look at the specification in the interface:
```
Returns {@code true} if the iteration has more elements.
```

**Review by garydgregory — CHANGES_REQUESTED** [review]
Hello @alexey-pelykh 
- You are missing a unit test for Iterator.forEachRemaining()
- Also edge cases like what if the first list, a middle list or the last list is empty.
- IOW what is the expected behavior if in your example `letters`, `numbers`, or `symbols` is empty
- What should happen if the lists point to each other or have cycles?

**Comment by garydgregory** [comment]
Just FTR https://stackoverflow.com/questions/32131987/how-can-i-make-cartesian-product-with-java-8-streams

File: src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java (hunk 17 above)
Comment (alexey-pelykh) [reply src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java:155]: @garydgregory reasonable, yet that would fall out of line with other iterators present in the library - most of them use `remove() is not supported` or more detailed `remove() method is not supported for an ObjectArrayIterator`. Taking that into account, is this change request remains and is a blocking one?

File: src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java (hunk 15 above)
Comment (alexey-pelykh) [reply src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java:89]: I agree that `@SafeVarargs` should be here, yet I don't get it why do you refer to the null-guards as the reason since it's not related with Object[] arriving in real-time. However, I did address that anyways so the annotation should be applied.

File: src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java (hunk 14 above)
Comment (alexey-pelykh) [reply src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java:112]: While I agree that it's less verbose, I don't think that justifies all the call chain and objects creation that need to happen in that line. Is this change request a blocking one?

**Comment by alexey-pelykh** [comment]
Hi @garydgregory:

> You are missing a unit test for Iterator.forEachRemaining()

What is the rationale to add that given that `forEachRemaining` is a `hasNext()`/`next()` wrapper and those are explicitly tested?

> Also edge cases like what if the first list, a middle list or the last list is empty.

Reasonable, added.

> IOW what is the expected behavior if in your example letters, numbers, or symbols is empty

Added tests to be explicit. If at least one list is empty, cartesian product returns no tuples.

> What should happen if the lists point to each other or have cycles?

If any of the passed iterables has cycle and thus is infinite, the iterator would also produce an infinite result. Here the behavior is the same as in nested for-loop. Could you please elaborate on the lists pointing to each other case as I'm not exactly sure I got you there.

File: src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java (hunk 18 above)
Comment (alexey-pelykh) [reply src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java:97]: Fair enough, not my preferred form yet no objection either.

File: src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java (hunk 17 above)
Comment (garydgregory) [reply src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java:155]: Hello @alexey-pelykh 
What I am saying is:
- Change the exception message
- Add a test that asserts the exception is always thrown

File: src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java (hunk 14 above)
Comment (garydgregory) [reply src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java:112]: The difference (to me) is whether you make the change or I do it, which affects how fast I can process PRs and requests in general. If a PR is more work for me, I just put it aside and get back to it, later. If I took the time to review the PR, test it, and write the comment, then I expect the change to be made.

**Comment by garydgregory** [comment]
> Hi @garydgregory:
> 
> > You are missing a unit test for Iterator.forEachRemaining()
> 
> What is the rationale to add that given that `forEachRemaining` is a `hasNext()`/`next()` wrapper and those are explicitly tested?
> 
> > Also edge cases like what if the first list, a middle list or the last list is empty.
> 
> Reasonable, added.
> 
> > IOW what is the expected behavior if in your example letters, numbers, or symbols is empty
> 
> Added tests to be explicit. If at least one list is empty, cartesian product returns no tuples.
> 
> > What should happen if the lists point to each other or have cycles?
> 
> If any of the passed iterables has cycle and thus is infinite, the iterator would also produce an infinite result. Here the behavior is the same as in nested for-loop. Could you please elaborate on the lists pointing to each other case as I'm not exactly sure I got you there.

If the API allows an infinite loop, this is what will happen: Someone will send an email to our security mailing list saying they found a critical DOS vulnerability and ask for a CVE to be credited to them; time will be taken analyzing, deciding what to do and replying to the poster. This is why we've fixed these types of bugs overtime and added tests like 
- `HashCodeBuilderTest.testReflectionObjectCycle()`
- `ToStringBuilderTest.testSelfInstanceTwoVarsReflectionObjectCycle()`
- `ToStringBuilderTest.testSelfInstanceVarReflectionObjectCycle()`
- `ToStringBuilderTest.testSimpleReflectionObjectCycle()`
- `StrSubstitutorTest.testCyclicReplacement()`
- `ToStringBuilderTest.testReflectionArrayAndObjectCycle()`
- `ToStringBuilderTest.testReflectionArrayArrayCycle()`
- ...

And yes, there are probably more of these types of bugs lurking around, so let's not add new ones ;-)

File: src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java (hunk 14 above)
Comment (alexey-pelykh) [reply src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java]: That is what I meant as "blocking" - as in, the reviewer is not open to discuss the change request if the author has counter-arguments towards it.

Change done.

File: src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java (hunk 17 above)
Comment (alexey-pelykh) [reply src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java]: Thanks for clarifying, change done.

**Comment by alexey-pelykh** [comment]
> And yes, there are probably more of these types of bugs lurking around, so let's not add new ones ;-)

Absolutely not my intent! That said, I guess I'm too dumb to understand how that bug may be exploited in this case.

- If a `CartesianProductIterator` is passed an infinite iterable, the result is expectedly infinite - the same was as for loop over infinite iterable is going to be infinite.
- If a `CartesianProductIterator` is passed an iterable that is modified while iterating over the iterator, it may or may not yield into an infinite iterator.
- If a `CartesianProductIterator` is passed a fake iterable that always returns reference to the `CartesianProductIterator` it was passed to ("iterating over self") - it may be infinite.

However, the infinite loop won't happen in the`CartesianProductIterator` itself, but the `hasNext()` will always be true - so iterating over such an iterator will be infinite, alike iterating over a "random generator" iterator.

**Comment by garydgregory** [comment]
> @garydgregory would it be possible to restart 23-ae? It's not exactly clear what is wrong

EA builds are marked experimental in the CI and fail when the tooling has not been adapted Java EA versions. In this case probably some locale changes.

**Comment by garydgregory** [comment]
> > You are missing a unit test for Iterator.forEachRemaining()
> 
> What is the rationale to add that given that `forEachRemaining` is a `hasNext()`/`next()` wrapper and those are explicitly tested?
> 

The new class implements an interface. The tests should show what are the expectations for all interface methods, the ones that are actually implemented by the new class _and_ the behavior that is inherited from the default method.

There is no guarantee that implementing the methods that a default interface calls is actually correctly implementing the contract for the default method. Testing the default methods shows that it works as expected, regardless of whether the default method is implemented or inherited. You are also showing as a PR author that you've considered the big picture and are thorough, all the pieces work individually and together, as expected.

**Comment by alexey-pelykh** [comment]
> Testing the default methods shows that it works as expected, regardless of whether the default method is implemented or inherited.

👍  test added

**Comment by garydgregory** [comment]
@alexey-pelykh 

Thank you for your PR and patience. Merged! 🚀
